server/src/main/java/org/tctalent/server/service/db/impl/CandidateServiceImpl.java

```diff
@@ -405,2392 +405,2410 @@ public Page<Candidate> searchCandidates(CandidateEmailOrPhoneSearchRequest reque
+
+    SurveyType surveyType = null;
+    if (request.getSurveyTypeId() != null) {
+      // Load the education level from the database - throw an exception if not found
+      surveyType = surveyTypeRepository.findById(request.getSurveyTypeId())
+          .orElseThrow(() -> new NoSuchObjectException(EducationLevel.class, request.getSurveyTypeId()));
+    }
+    candidate.setSurveyType(surveyType);
+    candidate.setSurveyComment(request.getSurveyComment());
+    return save(candidate, true);
+  }
+
+  @Override
+  public Candidate updateCandidateMedia(long id, UpdateCandidateMediaRequest request) {
+    User loggedInUser = authService.getLoggedInUser()
+        .orElseThrow(() -> new InvalidSessionException("Not logged in"));
+
+    Set<Country> sourceCountries = userService.getDefaultSourceCountries(loggedInUser);
+    Candidate candidate = this.candidateRepository.findByIdLoadUser(id, sourceCountries)
+        .orElseThrow(() -> new NoSuchObjectException(Candidate.class, id));
+
+    candidate.setMediaWillingness(request.getMediaWillingness());
+    return save(candidate, true);
+  }
+
+  @Override
+  public Candidate updateCandidateRegistration(long id, UpdateCandidateRegistrationRequest request) {
+    User loggedInUser = authService.getLoggedInUser()
+        .orElseThrow(() -> new InvalidSessionException("Not logged in"));
+
+    Set<Country> sourceCountries = userService.getDefaultSourceCountries(loggedInUser);
+    Candidate candidate = this.candidateRepository.findByIdLoadUser(id, sourceCountries)
+        .orElseThrow(() -> new NoSuchObjectException(Candidate.class, id));
+
+    candidate.setExternalId(request.getExternalId());
+    candidate.setExternalIdSource(request.getExternalIdSource());
+    candidate.setPartnerRef(request.getPartnerRef());
+    candidate.setUnhcrStatus(request.getUnhcrStatus());
+    candidate.setUnhcrConsent(request.getUnhcrConsent());
+    candidate.setUnhcrNumber(request.getUnhcrNumber());
+    candidate.setUnrwaRegistered(request.getUnrwaRegistered());
+    candidate.setUnrwaNumber(request.getUnrwaNumber());
+    return save(candidate, true);
+  }
+
+  @Override
+  @Transactional
+  public boolean deleteCandidate(long id) {
+    Candidate candidate = candidateRepository.findById(id).orElse(null);
+    if (candidate != null) {
+      String textSearchId = candidate.getTextSearchId();
+      candidateRepository.delete(candidate);
+      if (textSearchId != null) {
+        candidateEsRepository.deleteById(textSearchId);
+      }
+      return true;
+    } else {
+      return false;
     }
+  }
 
-    @Override
-    public @NonNull Candidate getCandidate(long id) throws NoSuchObjectException {
-        final Candidate candidate = candidateRepository.findById(id)
-            .orElseThrow(() -> new NoSuchObjectException(Candidate.class, id));
+  @Override
+  @Transactional(rollbackFor = Exception.class)
+  public LoginRequest register(RegisterCandidateRequest request, HttpServletRequest httpRequest) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 980-1024 into a separate method.